### PR TITLE
Expose committed session cloning and build prefix reuse above it

### PR DESCRIPTION
**Pull request by mergennachin**
Expose `Session::clone_async(upto)` for independently writable committed prefixes. A caller-owned `PrefixCache` handles token matching and LRU eviction; `capture_prompt(...).wrap(...)` requests the prompt clone before the first user callback, and caller-thread `collect()` inserts it after generation. Capture binds to session identity and survives moves without retaining source ownership. Lookup tries each snapshot only at its longest matching boundary. The MLX example uses this helper and forwards the final prompt token for fresh logits. The cache factory API stays unchanged; session limits are checked through the constructed cache's `BatchControl::max_seqs()`.

Review the Session/Runner lifecycle first, then ModuleExecutor/backend clone contracts, prefix policy, and MLX integration.

**214 shared native tests pass on Apple M1 Pro**, including 14 capture tests for timing, reuse, failures, handle lifetime, and moved/reassigned/destroyed sources. All five native targets and the standalone MLX batched runner build; lint passes. GPU execution was not repeated for this lifetime change; the following GPU/model results are from `ddf92b40d1`.

All 49 MLX tests pass. Gemma 3 1B validation covers BF16/FP16, full/selected logits, both layouts, physical ring wrap, branches, closure, and concurrency: 16 smoke groups, 88 cache/work checks, and 44 cold export comparisons pass. Warm repeats forward 1 of 23/1280 prompt tokens. **Normal cold/warm initial logits exceed tolerance in 72/88 cases.** The public header recommends opt-in greedy experiments until non-greedy sampling is validated; even greedy token parity is not guaranteed. Clone controls with matching schedule/geometry are byte-exact. A 23-token control also gives byte-exact cold/split/replay/clone logits in both dtypes with one-token prefill. Isolated probes identify shape-dependent MLX projection and attention arithmetic; precision promotions did not resolve the normal cold/warm gap. The four existing exports were reused and hash-verified.

Authored with OpenAI Codex.

### CODE REVIEW

**Comment by metascroy** [comment]
Could we expose cloning as a general session operation and build prefix caching on top of that, rather than adding reuse_prefix() and cache_prefix() to Executor?

Concretely, Session::clone_async(upto) would route through the runner to Executor::clone(source, upto), with ModuleExecutor delegating to BatchControl::seq_clone(). The result would be a new independently writable session containing the committed prefix, with backend retention checks and resource limits determining whether cloning succeeds.

The prefix cache would then own retained snapshot sessions and the token-matching/LRU policy. A lookup would find the longest usable match and return a clone plus the matched-token count, leaving at least the final prompt token to be forwarded. On a miss, the caller opens a fresh session; on eviction, the cache releases its retained session.

Thoughts?

File: extension/llm/batching/module_executor.cpp [hunk 1]
```diff
@@ -284,11 +289,15 @@ Result<std::unique_ptr<ModuleExecutor>> ModuleExecutor::create(
...
+  const std::int64_t reserved_sequences =
+      static_cast<std::int64_t>(max_sessions) + prefix_cache_entries;
+  const std::int64_t required_ids =
+      reserved_sequences + (prefix_cache_entries > 0 ? 1 : 0);
+  if (required_ids > std::numeric_limits<int>::max() / max_session_tokens) {
     ET_LOG(Error, "ModuleExecutor: total cache capacity exceeds int range");
     return Error::InvalidArgument;
   }
-  cfg->capacity = max_sessions * max_session_tokens;
+  cfg->capacity = static_cast<int>(required_ids) * max_session_tokens;
```
[the remark this answers is not in the document]
Comment (mergennachin) [reply extension/llm/batching/module_executor.cpp]: Updated following [the API review](https://github.com/pytorch/executorch/pull/22812#discussion_r4019050819): `816bd2fb28` removes the separately registered sequence-limit metadata and restores the existing post-construction `BatchControl::max_seqs()` check. The cell layout remains the source of its limit. This supersedes the pre-allocation fix described in my earlier reply.

Very large requests can still allocate CPU cell metadata before rejection; GPU KV buffers are allocated lazily. Pre-construction capability discovery is deferred and should come from the builder itself. The two pre-construction-only tests were removed. All 200 remaining shared tests pass, all five native targets build, and lint passes; GPU execution was not repeated for this revision.

File: extension/llm/batching/test/runner_test.cpp [hunk 2]
```diff
@@ -135,6 +136,146 @@ struct Fixture {
...
+      history_.erase(sid);
+    }
+    FakeExecutor::close_session(sid);
+  }
+
+  std::optional<SessionId> clone(SessionId source, Position upto) override {
+    {
+      std::lock_guard<std::mutex> lock(state_mutex_);
+      clones_.push_back(
+          {source, upto, seen().size(), std::this_thread::get_id()});
```
[the remark this answers is not in the document]
Comment (mergennachin) [reply extension/llm/batching/test/runner_test.cpp:169]: These are distinct mutexes: [CloneExecutor::clone()](https://github.com/pytorch/executorch/blob/a2e5e939fada57969a6c63fbcaa2747b7c59b171/extension/llm/batching/test/runner_test.cpp#L165) holds the derived class's `state_mutex_`, while inherited [FakeExecutor::seen()](https://github.com/pytorch/executorch/blob/a2e5e939fada57969a6c63fbcaa2747b7c59b171/extension/llm/batching/test/fake_executor.h#L177) locks the base class's `mutex_`. There is no recursive acquisition of the same mutex. All 150 batching tests passed, and three relevant clone tests passed 20 shuffled repetitions each (60/60), including a successful committed-prefix clone and callback-before-next-forward ordering.

File: backends/mlx/examples/llm/run_llm_batched.cpp [hunk 3]
```diff
@@ -202,10 +241,25 @@ void submit_prompt(
     config.stop_tokens = stop_tokens;
     config.seed = FLAGS_seed;
 
-    result.handle = session.generate_async(
-        std::move(*encoded),
+    std::vector<batching::Token> suffix(
+        result.prompt_tokens.begin() + result.cached_prompt_tokens,
+        result.prompt_tokens.end());
+    result.handle = result.session->generate_async(
+        std::move(suffix),
```
[the remark this answers is not in the document]
Comment (mergennachin) [reply backends/mlx/examples/llm/run_llm_batched.cpp:250]: The numerical investigation at `ddf92b40d1` now isolates MLX arithmetic differences in the 23-token BF16/FP16 selected-logits, batched-sequence reproducer. All traced/untraced and archived output checks pass byte-for-byte. The first difference is Q projection (op 35) on identical input and weights. Isolated attention replay reproduces the larger first-layer discrepancy: rounding QK scores and softmax probabilities to the model dtype reconstructs every unfused output value in both dtypes. Matched clone/original replay remains byte-exact.

A native control processing the entire prompt one token per forward gives byte-identical cold, split, original replay, and clone replay logits: all 12 pairwise comparisons pass across BF16/FP16, over 262,144 logits per row. All 32 main/preparation schedules and cache/work counters are checked, including reuse of 22 tokens and fresh execution at position 22. Ordinary multi-token prefill still reproduces the original 0.75 BF16 / 0.078125 FP16 maximum error. This supports schedule-dependent numerical drift; no prefix-policy or clone-state defect is demonstrated by this repro.

Uniform low-precision attention, FP32 attention, and FP32 attention plus projections were also tested; none resolves the ordinary cold/warm tolerance failure. This does not invalidate the local attention diagnosis, but full-model error growth is not completely traced. Serial prefill is a diagnostic control, not a production workaround.

The broader validation remains 251 native tests, 16 smoke groups, 88 cache/work checks, and 44 cold full/selected comparisons passing, with 72/88 ordinary cold/warm initial-logit comparisons outside the unchanged tolerances. The matching-schedule result covers one short prompt and one initial prediction; it is not a general numerical-equivalence claim. Non-greedy sampling remains untested and prefix reuse remains disabled by default (`--prefix_cache_entries=0`). The PR summary now records these findings.

File: extension/llm/cache/cache_registry.h [hunk 4]
```diff
@@ -93,22 +94,33 @@ class ET_EXPERIMENTAL CacheFactory {
   CacheFactory() = default;
 
   // Registers one builder without replacing an existing entry. Returns
-  // InvalidArgument if builder is empty or the pair is already registered.
+  // InvalidArgument if builder is empty, max_seqs is non-positive, or the pair
+  // is already registered. A fixed sequence limit must be advertised here so
+  // requests exceeding it can be refused before the builder allocates storage.
   ET_NODISCARD Error register_builder(
       const std::string& backend_id,
       const std::string& kind,
```
Comment (metascroy) [inline extension/llm/cache/cache_registry.h:99]: Why do we need a separate max_seqs argument on registration? The cache already exposes this through BatchControl::max_seqs(), and the cell layout owns its limit. Can we retain the existing post-construction check for this change? If we need pre-construction capability discovery, that should come from the builder itself rather than separately supplied metadata.

Comment (mergennachin) [reply extension/llm/cache/cache_registry.h:99]: Agreed. The extra argument was intended to reject oversized requests before allocation, but it duplicates the layout's limit. Removed it in `816bd2fb28` and restored the existing factory API and post-construction `BatchControl::max_seqs()` check. The MLX registration test now checks the limit on the constructed cache. Future pre-construction discovery should come from the builder itself.

Validation: 200 shared tests pass; all five native targets build; lint passes. GPU execution was not repeated for this revision.

File: backends/mlx/examples/llm/run_llm_batched.cpp [hunk 5]
```diff
@@ -202,10 +241,25 @@ void submit_prompt(
...
         std::move(config),
-        [emitter](const batching::GenerationUpdate& update) {
+        [emitter, retain_prefix, &result](
+            const batching::GenerationUpdate& update) {
```
Comment (metascroy) [inline backends/mlx/examples/llm/run_llm_batched.cpp]: Since correct snapshot capture depends on requesting a clone from the first output callback, could PrefixCache provide a small helper for that pattern? Ideally it would wrap the generation callback, request the clone once without blocking, and expose a caller-thread collection/insertion step. That would keep users from having to reproduce the timing, future ownership, and failure-handling logic in each integration.

**Review by metascroy — APPROVED** [review]

File: backends/mlx/examples/llm/run_llm_batched.cpp (hunk 5 above)
Comment (mergennachin) [reply backends/mlx/examples/llm/run_llm_batched.cpp]: Implemented in `59d22ca57d`:

```cpp
auto capture = prefix_cache.capture_prompt(session, prompt_tokens);
auto generation = session.generate_async(
    suffix, config, capture.wrap(on_update));
generation.wait();
capture.collect();
```

The wrapper requests one clone before the first nonempty user callback, without waiting. Collection inserts best-effort on the caller thread. Capture binds to the original session ID and status, so moving or reassigning the source wrapper is safe; it does not own the source or delay closure. The MLX example and docs use this flow.

All 214 shared tests pass, including 14 capture cases and three source-lifetime regressions. All five native targets and the standalone MLX runner build; lint passes. GPU execution was not repeated. The header also states lookup's one-boundary-per-snapshot behavior and the MLX cold/warm parity caveat, including the lack of a greedy-token guarantee.

**Comment by mergennachin** [comment]
@claude Review this PR

**Comment by mergennachin** [comment]
@claude Re-review now
